File: AppBox/ViewController/PreferencesViewController/ThirdPartyPreferencesViewController/ThirdPartyPreferencesViewController.m
//
//  ThirdPartyPreferencesViewController.m
//  AppBox
//
//  Created by Vineet Choudhary on 10/05/17.
//  Copyright © 2017 Developer Insider. All rights reserved.
//

#import "ThirdPartyPreferencesViewController.h"

@implementation ThirdPartyPreferencesViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [slackChannelTextField setStringValue:[UserData userSlackChannel]];
    [microsoftTeamWebHook setStringValue:[UserData userMicrosoftTeamWebHook]];
    [slackMessageTextField setStringValue:[UserData userSlackMessage]];
}


- (BOOL)validateWebHookURLWithSlack:(BOOL)slack msTeams:(BOOL)msTeams{
    NSString *slackURL = [slackChannelTextField.stringValue stringByRemovingPercentEncoding];
    if ([NSURL URLWithString:slackURL] == nil && slack) {
        [Common showAlertWithTitle:@"Error" andMessage:@"Please enter a valid Slack WebHook URL."];
        return NO;
    }
    
    NSString *microsoftTeamURL = [microsoftTeamWebHook.stringValue stringByRemovingPercentEncoding];
    if ([NSURL URLWithString:microsoftTeamURL] == nil && msTeams) {
        [Common showAlertWithTitle:@"Error" andMessage:@"Please enter a valid Microsoft Team WebHook URL."];
        return NO;
    }
    return YES;
}

- (IBAction)saveButtonTapped:(NSButton *)sender {
    [self.view.window makeFirstResponder:self.view];
    if (![self validateWebHookURLWithSlack:YES msTeams:YES]) {
        return;
    }
    [UserData setUserSlackChannel:slackChannelTextField.stringValue];
    [UserData setUserMicrosoftTeamWebHook:microsoftTeamWebHook.stringValue];
    [UserData setUserSlackMessage:slackMessageTextField.stringValue];
    [ABHudViewController showStatus:@"Details Saved!" forSuccess:YES onView:self.view];
}

- (IBAction)testSlackButtonTapped:(NSButton *)sender {
    [self.view.window makeFirstResponder:self.view];
    if (![self validateWebHookURLWithSlack:YES msTeams:NO]) {
        return;
    }
	NSString *webhook = [slackChannelTextField.stringValue stringByRemovingPercentEncoding];
	NSString *message = slackMessageTextField.stringValue;
    [ABHudViewController showStatus:@"Sending Test Message..." onView:self.view];
	[SlackClient sendMessage:[self demoIPAUploadInfo] webhook:webhook message:message completion:^(BOOL success) {
        [ABHudViewController hideAllHudFromView:self.view after:0];
        if (success) {
            [ABHudViewController showStatus:@"Message Sent." forSuccess:YES onView:self.view];
        } else {
            [ABHudViewController showStatus:@"Message Failed." forSuccess:NO onView:self.view];
        }
    }];
}

- (IBAction)testMicrosoftTeamButtonTapped:(NSButton *)sender {
    [self.view.window makeFirstResponder:self.view];
    if (![self validateWebHookURLWithSlack:NO msTeams:YES]) {
        return;
    }
	NSString *webhook = [microsoftTeamWebHook.stringValue stringByRemovingPercentEncoding];
	NSString *message = slackMessageTextField.stringValue;
    [ABHudViewController showStatus:@"Sending Test Message..." onView:self.view];
    [MSTeamsClient sendMessage:[self demoIPAUploadInfo] webhook:webhook message:message completion:^(BOOL success) {
        [ABHudViewController hideAllHudFromView:self.view after:0];
        if (success) {
            [ABHudViewController showStatus:@"Message Sent." forSuccess:YES onView:self.view];
        } else {
            [ABHudViewController showStatus:@"Message Failed." forSuccess:NO onView:self.view];
        }
    }];
}

- (IBAction)setupNewSlackWebhook:(NSButton *)sender {
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:abWebHookSetupURL]];
}

-(IPAUploadInfo *)demoIPAUploadInfo{
    //create a test ipa upload info for demo email
    IPAUploadInfo *ipaUploadInfo = [[IPAUploadInfo alloc] init];
    [ipaUploadInfo setName:@"TestApp"];
    [ipaUploadInfo setVersion:@"1.0"];
    [ipaUploadInfo setBuild:@"1"];
    [ipaUploadInfo setAppShortShareableURL:[NSURL URLWithString:@"https://getappbox.com"]];
    return ipaUploadInfo;
}

@end

File: AppBox/ViewController/PreferencesViewController/ThirdPartyPreferencesViewController/ThirdPartyPreferencesViewController.swift
//
//  ThirdPartyPreferencesViewController.swift
//  AppBox

import AppKit
import AppBoxCore
import os

public final class ThirdPartyPreferencesViewController: NSViewController {

    private static let log = Logger(subsystem: "com.developerinsider.AppBox", category: "ThirdPartyPreferences")
    private var model: ThirdPartyModel?

    public override func loadView() {
        let model = ThirdPartyModel(
			slackWebhook: UserData.userSlackChannel(),
			teamsWebhook: UserData.userMicrosoftTeamWebHook())
        self.model = model
        model.onSave = { [weak self] in self?.saveTapped() }
        model.onTestSlack = { [weak self] in self?.testSlackTapped() }
        model.onTestTeams = { [weak self] in self?.testTeamsTapped() }
        model.onSetupWebhook = {
            if let url = URL(string: "https://my.slack.com/apps/new/A0F7XDUAZ-incoming-webhooks") {
                NSWorkspace.shared.open(url)
            }
        }
        view = ThirdPartyPreferencesHost.makeView(model: model)
    }

    // MARK: - Actions (invoked by the SwiftUI form's callbacks)

    private func isValidWebHookURL(_ urlString: String, name: String, allowEmpty: Bool = false) -> Bool {
        let decoded = urlString.removingPercentEncoding ?? urlString
        if decoded.isEmpty {
            if allowEmpty { return true }
            Common.showAlert(withTitle: "Error", andMessage: "Please enter a \(name) WebHook URL.")
            return false
        }
        if URL(string: decoded) == nil {
            Common.showAlert(withTitle: "Error", andMessage: "Please enter a valid \(name) WebHook URL.")
            return false
        }
        return true
    }

    private func saveTapped() {
        guard let model else { return }
        let slack = model.slackWebhookText
        let teams = model.teamsWebhookText
        guard isValidWebHookURL(slack, name: "Slack", allowEmpty: true),
              isValidWebHookURL(teams, name: "Microsoft Team", allowEmpty: true) else { return }
        UserData.setUserSlackChannel(slack)
        UserData.setUserMicrosoftTeamWebHook(teams)
        Common.showAlert(withTitle: "Saved", andMessage: "Integration settings saved.")
    }

	private func testSlackTapped() {
		guard let model else { return }
		let webhook = model.slackWebhookText
		guard isValidWebHookURL(webhook, name: "Slack") else { return }
		guard requireDropboxSession() else { return }
		model.setTestingSlack(true)
		SlackClient.sendMessage(
			demoIPAUploadInfo(),
			webhook: webhook.removingPercentEncoding ?? webhook) { [weak self] success in
				DispatchQueue.main.async {
					self?.model?.setTestingSlack(false)
					self?.finishWebhookTest(success, service: "Slack")
				}
			}
	}

    private func testTeamsTapped() {
        guard let model else { return }
        let webhook = model.teamsWebhookText
        guard isValidWebHookURL(webhook, name: "Microsoft Team") else { return }
        guard requireDropboxSession() else { return }
        model.setTestingTeams(true)
        MSTeamsClient.sendMessage(
			demoIPAUploadInfo(),
			webhook: webhook.removingPercentEncoding ?? webhook) { [weak self] success in
				DispatchQueue.main.async {
					self?.model?.setTestingTeams(false)
					self?.finishWebhookTest(success, service: "Microsoft Teams")
				}
			}
    }

    private func finishWebhookTest(_ success: Bool, service: String) {
        if success {
            Common.showAlert(
				withTitle: "Message Sent",
				andMessage: "A test message was sent to \(service).")
        } else {
            Self.log.error("\(service, privacy: .public) test webhook failed.")
            Common.showAlert(
				withTitle: "\(service) Test Failed",
				andMessage: "Couldn't send a test message to \(service). Please check the webhook URL.")
        }
    }

    /// Test notifications go through the backend, which authenticates the caller by their Dropbox session — so a test needs a connected account (previously the send was direct from the app).
    private func requireDropboxSession() -> Bool {
        guard DropboxSession.isAuthorized else {
            Common.showAlert(
				withTitle: "Connect Dropbox",
				andMessage: "Connect your Dropbox account first to send a test notification.")
            return false
        }
        return true
    }

    private func demoIPAUploadInfo() -> IPAUploadInfo {
        let info = IPAUploadInfo()
        info.name = "TestApp"
        info.version = "1.0"
        info.build = "1"
        info.appShortShareableURL = URL(string: "https://getappbox.com")
        return info
    }
}
